### PR TITLE
[Prototype] exceptions stracktrace recovery

### LINKED ISSUE #493: Basic stacktrace augmentation
After #333 we can do a basic [pluggable? opt-in or opt-out?] stacktrace augmentation on `throw` call-site.

The basic idea is to attempt to reflectively instantiate exception of the same type at `throw` call-site with original exception as a cause.

For example currently this code
```
val deferred = async {
  throw IllegalStateException("Some invariant failed")
}

try {
  deferred.await()
} catch (t: Throwable) {
  t.printStackTrace()
}
```

will print
```
java.lang.IllegalStateException: Some invariant failed
  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1$deferred$1.doResume(AsyncJvmTest. 
 kt:15)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)
```
after stacktrace augmentation it will be

```
java.lang.IllegalStateException: java.lang.IllegalStateException: Some invariant failed

  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1.doResume(AsyncJvmTest.kt:21)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)
  at kotlinx.coroutines.experimental.DispatchedContinuation.run(Dispatched.kt:13)
  at kotlinx.coroutines.experimental.EventLoopBase.processNextEvent(EventLoop.kt:132)
  ....
  at kotlinx.coroutines.experimental.AsyncJvmTest.simpleTest(AsyncJvmTest.kt:13) // <- WHERE EXCEPTION IS ACTUALLY THROWN
Caused by: java.lang.IllegalStateException: Some invariant failed
  at kotlinx.coroutines.experimental.AsyncJvmTest$simpleTest$1$deferred$1.doResume(AsyncJvmTest.kt:15)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)

```

FJP for FJ-tasks does exactly the same: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/ForkJoinTask.java#l582

### PR DESCRIPTION
**Pull request by qwwdfsad**
1) Provide entry-points for stack-trace recovery at the call-site (not at suspension point), which can be improved in 1.3
2) First-class support in `await` and channel
3) Enable only in debug mode

Things to do
1) Discuss whether we want to hide this behavior under another system property
2) Implement caching on JDK 1.7 with ClassValueMap (we can't use CHM because it will be a memory leak)
3) Implement `withContext` support

Fixes #493 and (partially) #74

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
This would be amazing! I currently do something very similar by injecting the callsite stacktrace as a fake cause of the suspension point exception.

File: core/kotlinx-coroutines-core/src/internal/Exceptions.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                break
+            }
+        }
+    } catch (e: Exception) {
+        // Do nothing
+    }
+    return newException
+}
+
+private fun fillInStackTrace(continuation: Continuation<*>): ArrayList<StackTraceElement> {
```
Comment (elizarov) [inline core/kotlinx-coroutines-core/src/internal/Exceptions.kt]: It does not "fill" anything, so it should be called `getStackTrace`.

File: core/kotlinx-coroutines-core/src/internal/Exceptions.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    var last = continuation
+    while (true) {
+        last = getCompletion(last) ?: break
+        stack.add(stackTraceElement(last))
+    }
+    return stack
+}
+
+// TODO basic stub before 1.3
+private fun stackTraceElement(continuation: Continuation<*>): StackTraceElement {
```
Comment (elizarov) [inline core/kotlinx-coroutines-core/src/internal/Exceptions.kt]: Should return a nullable result. An object may not have a stack trace itself, but may be a completion. An good example here is `CancellableContinuationImpl` that should be supported.

File: core/kotlinx-coroutines-core/src/internal/Exceptions.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+internal actual fun <E : Throwable> recoverStackTrace(exception: E): E {
+    if (!DEBUG || exception is CancellationException) {
+        return exception
+    }
+
+    return tryWrapException(exception) ?: exception
+}
+
+internal actual fun <E : Throwable> recoverStackTrace(exception: E, continuation: Continuation<*>): E {
+    if (!DEBUG || exception is CancellationException || continuation !is CoroutineImpl) {
```
Comment (elizarov) [inline core/kotlinx-coroutines-core/src/internal/Exceptions.kt]: `continuation !is CoroutineImpl` is too restrictive for experimental coroutines. `CancellableContinuationImpl` shall be supported, too. It is needed for basic example that use `Deferred.await()`, for example.

**Pull request closed without merging**